zebra/dplane_fpm_pb.c

```diff
@@ -0,0 +1,973 @@
+      "Remote FPM server port\n")
+{
+	union sockunion su;
+	uint16_t fpm_port = 0;
+
+	if (argc == 5)
+		fpm_port = strtol(argv[4]->arg, NULL, 10);
+
+	if (str2sockunion(argv[2]->arg, &su) == -1) {
+		vty_out(vty, "%% Invalid address: %s\n", argv[2]->arg);
+		return CMD_WARNING;
+	}
+
+	/* Assign gfpc->addr */
+	if (su.sa.sa_family == AF_INET) {
+		su.sin.sin_port =
+		   fpm_port ? htons(fpm_port) : htons(SOUTHBOUND_DEFAULT_PORT);
+		memcpy(&gfpc->addr, &su.sin, sizeof(su.sin));
+	}
+
+	if (su.sa.sa_family == AF_INET6) {
+		su.sin6.sin6_port =
+		   fpm_port ? htons(fpm_port) : htons(SOUTHBOUND_DEFAULT_PORT);
+		memcpy(&gfpc->addr, &su.sin6, sizeof(su.sin6));
+	}
+
+	event_add_event(gfpc->fthread->master, fpm_process_event, gfpc,
+			FPE_RECONNECT, &gfpc->t_event);
+	return CMD_SUCCESS;
+}
+
+DEFPY(no_fpm_set_address, no_fpm_set_address_cmd,
+      "no fpm address [<A.B.C.D|X:X::X:X> [port <1-65535>]]",
+      NO_STR
+      FPM_STR
+      "FPM remote listening server address\n"
+      "Remote IPv4 FPM server\n"
+      "Remote IPv6 FPM server\n"
+      "FPM remote listening server port\n"
+      "Remote FPM server port\n")
+{
+	event_add_event(gfpc->fthread->master, fpm_process_event, gfpc,
+			FPE_DISABLE, &gfpc->t_event);
+	return CMD_SUCCESS;
+}
+
+
+DEFPY(fpm_show_counters, fpm_show_counters_cmd,
+      "show fpm counters",
+      SHOW_STR
+      FPM_STR
+      "FPM statistic counters\n")
+{
+	vty_out(vty, "%30s\n%30s\n", "FPM counters", "============");
+
+#define SHOW_COUNTER(label, counter)                                           \
+	vty_out(vty, "%28s: %u\n", (label), (counter))
+
+	SHOW_COUNTER("Input bytes", gfpc->counters.bytes_read);
+	SHOW_COUNTER("Output bytes", gfpc->counters.bytes_sent);
+	SHOW_COUNTER("Output buffer current size", gfpc->counters.obuf_bytes);
+	SHOW_COUNTER("Output buffer peak size", gfpc->counters.obuf_peak);
+	SHOW_COUNTER("Connection closes", gfpc->counters.connection_closes);
+	SHOW_COUNTER("Connection errors", gfpc->counters.connection_errors);
+	SHOW_COUNTER("Data plane items processed",
+		     gfpc->counters.dplane_contexts);
+	SHOW_COUNTER("Data plane items enqueued", gfpc->counters.ctxqueue_len);
+	SHOW_COUNTER("Data plane items queue peak",
+		     gfpc->counters.ctxqueue_len_peak);
+	SHOW_COUNTER("Buffer full hits", gfpc->counters.buffer_full);
+	SHOW_COUNTER("User FPM configurations", gfpc->counters.user_configures);
+	SHOW_COUNTER("User FPM disable requests", gfpc->counters.user_disables);
+
+#undef SHOW_COUNTER
+
+	return CMD_SUCCESS;
+}
+
+DEFPY(fpm_reset_counters, fpm_reset_counters_cmd,
+      "clear fpm counters",
+      CLEAR_STR
+	  FPM_STR
+	  "FPM statistic counters\n")
+{
+	event_add_event(gfpc->fthread->master, fpm_process_event, gfpc,
+			FPE_RESET_COUNTERS, &gfpc->t_event);
+	return CMD_SUCCESS;
+}
+
+static int fpm_write_config(struct vty *vty)
+{
+	struct sockaddr_in *sin;
+	struct sockaddr_in6 *sin6;
+	int written = 0;
+
+	if (gfpc->disabled)
```

Comment:
so I think this config is a little unclear. there's a difference between config (including the default config you have compiled in) and whether the config is active, enabled, in-use. you want to show the config if it's present, whether or not the state is 'enabled'. and maybe you should adjust your 'show' command to be more general - not just showing the statistics, but also showing the current address+port, and the status?

Reply:
@mjstapp 

We don't have an explicit enable/disable command.

`gfpc->disabled` is set to false when the config is present
`gfpc->disabled` is set to true when the config is not present

For this reason, we show the config only when `gfpc->disabled` is false.

This is the same approach used in `dplane_fpm_nl`.

Are you suggesting a different approach here?